Refactor `expDesign`

marcobarilari opened the pull request:
This refers to the function in use for MT+ localizer (nb: MT/MST has its own design function for the moment)

TO DO:

### Refactoring
- [x] make it "stand alone"  #69 
- [x] make vars name transparent #67 
- [x] make fixation target set a separate function
- [x] modularize the function for "task" (considering to make them part of CPP_PTB)

### Features improvements
- [ ] ~~make it flexible in case someone wants only one condition (either `static` or `motion` or more than two conditions)~~ moved to #88 
- [ ] ~~add a task for different speed targets (either `faster` or `slower`), might be better to develop this in another or to easy the weight of this one. The ground is set though. Consider as the output of the function a matrix with 1 (is target) and 0 (no target) instead of the actual dot speed~~ moved to #89
- [ ] ~~improve and avoid `error('number of repetitions must be a multiple of max number of targets');`~~ moved to #90
- [x] make the target tasks optional (eg if I have two targets like fixation cross and speed, I set  `cfg.target.type = {'fixation_cross', 'speed'};`)
- [x] reconsider pseudorand. rule 3 in `setFixationTargets`:  `no more than ``nbRepetitions - 1`` target in the same event order`. Maybe better a rule of eg no less than 1 per each event position in the block (increase an equal distribution of the targets in the all event's position in the blocks across the run)

### Annex
- [x] improve doc in `README`
- [x] Improve doc in the function 
- [x] unit test

marcobarilari commented:
Unchecked to does will probably be moved to issues for future developments... this is already a big one sorry :)

Remi-Gau commented on subfun/expDesign.m:
```diff
@@ -59,69 +57,68 @@
 
     % Set variables here for a dummy test of this function
```
Will turn this into a test because this exactly how "smoke tests" work: 
feed some dummy input and make sure that the thing at least runs.

Remi-Gau commented on subfun/setSpeedTargets.m:
```diff
@@ -0,0 +1,23 @@
+function speeds = setSpeedTargets(cfg)
+
+    % Compute the matrix with the speed targets if requested, otherwise output will be only zeros
+    if sum(contains(cfg.target.type, 'speed')) ~= 0
```
won't touch this for now because at the moment either "side" of the if block does the same.

Remi-Gau commented:
added some tests.

do you want to also refactor the MT MST one ? It should be fairly straight forward now, no? (famous last words)

marcobarilari commented on subfun/design/setSpeedTargets.m:
```diff
@@ -0,0 +1,19 @@
+        % Outputs an "empty" matrix in case no speed task is required
+    else
+        [~, nbEventsPerBlock, ~, nbBlocks] = getDesignInput(cfg);
+        speeds = zeros(nbBlocks, nbEventsPerBlock) * cfg.dot.speedPixPerFrame;
```
Have not tested it yet but I think that having zeros might results in dots not moving, will update once tested

marcobarilari replied:
I confirm that zeros make the dots static

marcobarilari replied:
```suggestion
        speeds = ones(nbBlocks, nbEventsPerBlock) * cfg.dot.speedPixPerFrame;
```

marcobarilari commented on tests/test_setSpeedTargets.m:
```diff
@@ -0,0 +1,27 @@
+    assertEqual(speeds, ones(20, 12) * 28);
+
+    % try when the target are just for the fixation cross
+    cfg.target.type = {'fixation_cross'};
```
why fixation targets here?

Remi-Gau replied:
if there are only fixation cross as targets then `setSpeedTargets` must return only zeros.

So it is a second test.

Will break it down into 2 to avoid confusion

marcobarilari replied:
Ok here I have "updated" the test meaning that now, being WIP, the two tests are checking for the same results (a matrix full of ones).

Added a reminder in #89

marcobarilari commented:
Hey sorry for being late on the reply here.

> do you want to also refactor the MT MST one ? It should be fairly straight forward now, no? (famous last words)

I am tempted to do that in a new PR cause I suspect that a separate design for MT_MST might be useless now and with few 'if/switch' statements we can do the trick.

Also, I see that in some functions the copyright tag has been moved below the `function [out] = myFunc(in)` call. The style used 'everywhere' is to put it in the very first line, isn't it? 

I am a bit confused by the test for the `speed` task because I see code related to the `fixation` task there.  So my idea is to use `setSpeedTargets` to implement a task where the subj has to detect the eg 'faster dots trial'. The output matrix is full of values that are multiplicators of `cfg.dot.speed = 15;` so that if the matrix reports a `1`, then it has the set speed (no change in speed), if eg `2` for a given trial then that speed will be doubled. I know it is hard to get from the code as it is right now but it is because it is WIP and to implement when someone will need that. 
The idea is also the possibility to have >1 task in the same experiment if needed, as `fixation target` && `dot faster speed target`

Remi-Gau commented:
> Also, I see that in some functions the copyright tag has been moved below the `function [out] = myFunc(in)` call. The style used 'everywhere' is to put it in the very first line, isn't it?

Until recently miss_hit only allowed the top line, but that would mess up using the "help" command on Octave and very few matlab codebases put the copyright there.

So this has changed and we can now use the more conventional "copyright at the end of the help section".

No need to change all functions now, but if you run into a function where it is still in the top line, put it at the end of the help section.

Remi-Gau commented:
> I am tempted to do that in a new PR cause I suspect that a separate design for MT_MST might be useless now and with few 'if/switch' statements we can do the trick.

Agreed

Remi-Gau commented:
> I am a bit confused by the test for the `speed` task because I see code related to the `fixation` task there. So my idea is to use `setSpeedTargets` to implement a task where the subj has to detect the eg 'faster dots trial'. The output matrix is full of values that are multiplicators of `cfg.dot.speed = 15;` so that if the matrix reports a `1`, then it has the set speed (no change in speed), if eg `2` for a given trial then that speed will be doubled. I know it is hard to get from the code as it is right now but it is because it is WIP and to implement when someone will need that.

Ah OK I get it now: I was not sure what was expected so I sort of made a wild guess: then the code and test need to be updated.

Remi-Gau commented on subfun/design/setSpeedTargets.m:
```diff
@@ -0,0 +1,19 @@
+    %
+    % (C) Copyright 2021 CPP visual motion localizer developpers
+
+    if ismember('speed', cfg.target.type)
+        [~, nbEventsPerBlock, ~, nbBlocks] = getDesignInput(cfg);
+        speeds = ones(nbBlocks, nbEventsPerBlock) * cfg.dot.speedPixPerFrame;
+
+        % Outputs an "empty" matrix in case no speed task is required
+    else
+        [~, nbEventsPerBlock, ~, nbBlocks] = getDesignInput(cfg);
+        speeds = ones(nbBlocks, nbEventsPerBlock) * cfg.dot.speedPixPerFrame;
+
+    end
```
I know what confused me on this function.

At the moment both sides of the if block do exactly the same thing, so I thought either the If block is not necessary or the else part must be doing something different from the rest.

Do you see what I mean ?

marcobarilari replied:
Yes you are totally right, kind of impossible to guess. My bad.

Let's say that now the function has the structure for future developments. Let me add some "warnings" here and there to make sure it is clear

Remi-Gau replied:
just a couple of words below the [WIP] will do. :-)

Remi-Gau commented on subfun/design/expDesign.m:
```diff
@@ -0,0 +1,89 @@
+
+    % Set to 1 for a visualtion of the trials design order
+    if nargin < 2 || isempty(displayFigs)
+        displayFigs = 0;
```
no worries: I basically took all the mocked content and created the `getTestConfig` function with it.

We can just rename it ("get mock input"or something ) and move it into the design folder and make a call to that function in the expDesign function.

Makes sense?

marcobarilari replied:
Yes it makes sense, will also improve the doc related to this

marcobarilari commented:
I moved the left to does to new issues and brought back the mock configuration inside `expDesign`. Apologies if I push for this but it is a time saver for me and also might help future users to understand how it works and what are the inputs there (hope the new comments are clear).

Also added more info about the WIP situation of `setSpeedTargets` both in the readme and inside the function. Let me know if the test works after this or it needs a fix.

marcobarilari commented:
Merging this one
